nautobot/ipam/utils/migrations.py

```diff
@@ -0,0 +1,569 @@
+
+
+def get_next_cleanup_namespace(apps, vrf):
+    """
+    Try to get the next available Cleanup Namespace based on `vrf`.
+
+    Args:
+        apps: Django apps module
+        vrf (VRF): VRF instance
+
+    Returns:
+        Namespace
+    """
+    Namespace = apps.get_model("ipam", "Namespace")
+    VRF = apps.get_model("ipam", "VRF")
+
+    counter = 1
+    vrf_prefixes = vrf.prefixes.all()
+
+    global_ns = Namespace.objects.get(name="Global")
+    global_ns_prefixes = global_ns.prefixes.exclude(vrf=vrf)
+    global_dupe_prefixes = compare_duplicate_prefixes(vrf_prefixes, global_ns_prefixes)
+    global_dupe_vrfs = VRF.objects.filter(namespace=global_ns, name=vrf.name).exclude(pk=vrf.pk).exists()
+
+    if global_dupe_prefixes:
+        print(f"    VRF {vrf.name} has duplicate prefixes with NS {global_ns.name}")
+
+    if global_dupe_vrfs:
+        print(f"    VRF {vrf.name} has duplicate VRF name with NS {global_ns.name}")
+
+    if not any([global_dupe_prefixes, global_dupe_vrfs]):
+        return global_ns
+
+    # Iterate non-enforce_unique VRFS
+    # - Compare duplicate prefixes for each VRF
+    # - If a VRF has duplicates, it moves to a new namespace
+    while True:
+        base_name = f"{BASE_NAME} ({counter})"
+        namespaces = Namespace.objects.filter(name=base_name)
+
+        if not namespaces.exists():
+            return Namespace.objects.create(name=base_name, description="Created by Nautobot.")
+        namespace = namespaces.get()
+        ns_prefixes = namespace.prefixes.exclude(vrf=vrf)
+        dupe_prefixes = compare_duplicate_prefixes(vrf_prefixes, ns_prefixes)
+        dupe_vrfs = VRF.objects.filter(namespace=namespace, name=vrf.name).exclude(pk=vrf.pk).exists()
+
+        if dupe_prefixes:
+            print(f"    VRF {vrf.name} has duplicate prefixes with NS {namespace.name}")
+
+        if dupe_vrfs:
+            print(f"    VRF {vrf.name} has duplicate VRF name with NS {namespace.name}")
+
+        if any([dupe_prefixes, dupe_vrfs]):
+            counter += 1
+            continue
+
+        return namespace
+
+
+def get_next_prefix_namespace(apps, prefix, base_name=BASE_NAME):
+    Namespace = apps.get_model("ipam", "Namespace")
+
+    counter = 1
+    while True:
+        name = f"{base_name} ({counter})"
+        namespaces = Namespace.objects.filter(name=name)
+
+        if not namespaces.exists():
+            return Namespace.objects.create(name=name, description="Created by Nautobot.")
+        namespace = namespaces.get()
+        cidr = f"{prefix.network}/{prefix.prefix_length}"
+        has_dupe = namespace.prefixes.filter(network=prefix.network, prefix_length=prefix.prefix_length).exists()
+
+        if has_dupe:
+            print(f"    Prefix {cidr} is duplicated in NS {namespace.name}")
+            counter += 1
+            continue
+
+        return namespace
+
+
+def get_prefixes(qs):
+    return sorted(qs.values_list("network", "prefix_length"))
```

Comment:
Can we move this up closer to where it's used, and in general review the order functions are defined in this file? I had to jump around a lot in reviewing this file.